siatest/testgroup.go

```diff
@@ -0,0 +1,341 @@
+	if err := miner.MineBlock(); err != nil {
+		return nil, err
+	}
+	// Block until all hosts show up as active in the renters' hostdbs
+	if err := hostsInRenterDBCheck(miner, tg.renters, len(tg.hosts)); err != nil {
+		return nil, err
+	}
+	// Wait for all the renters to form contracts
+	if err := waitForContracts(miner, tg.renters, tg.hosts); err != nil {
+		return nil, err
+	}
+	// Make sure all nodes are synced
+	if err := synchronizationCheck(miner, tg.nodes); err != nil {
+		return nil, err
+	}
+	return tg, nil
+}
+
+// setRentersAllowance sets the allowance of each renter
+func setRentersAllowance(renters map[*TestNode]struct{}) error {
+	for renter := range renters {
+		if err := renter.RenterPost(defaultAllowance); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// NewGroupFromTemplate will create hosts, renters and miners according to the
+// settings in groupParams.
+func NewGroupFromTemplate(groupParams GroupParams) (*TestGroup, error) {
+	var params []node.NodeParams
+	// Create host params
+	for i := 0; i < groupParams.hosts; i++ {
+		params = append(params, node.Host(randomDir()))
+	}
+	// Create renter params
+	for i := 0; i < groupParams.renters; i++ {
+		params = append(params, node.Renter(randomDir()))
+	}
+	// Create miner params
+	for i := 0; i < groupParams.miners; i++ {
+		params = append(params, Miner(randomDir()))
+	}
+	return NewGroup(params...)
+}
+
+// randomDir is a helper functions that returns a random directory path
+func randomDir() string {
+	dir, err := TestDir(strconv.Itoa(fastrand.Intn(math.MaxInt32)))
+	if err != nil {
+		panic(build.ExtendErr("failed to create testing directory", err))
+	}
+	return dir
+}
+
+// synchronizationCheck makes sure that all the nodes are synced and follow the
+// same chain.
+func synchronizationCheck(miner *TestNode, nodes map[*TestNode]struct{}) error {
+	mcg, err := miner.ConsensusGet()
+	if err != nil {
+		return err
+	}
+	for node := range nodes {
+		err := Retry(100, 100*time.Millisecond, func() error {
+			ncg, err := node.ConsensusGet()
+			if err != nil {
+				return err
+			}
+			if mcg.CurrentBlock != ncg.CurrentBlock {
+				return errors.New("the node's current block doesn't equal the miner's")
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// mapToSlice converts a map of TestNodes into a slice
+func mapToSlice(m map[*TestNode]struct{}) []*TestNode {
+	tns := make([]*TestNode, 0, len(m))
+	for tn := range m {
+		tns = append(tns, tn)
+	}
+	return tns
+}
+
+// waitForContracts waits until the renters have formed contracts with the
+// hosts in the group.
+func waitForContracts(miner *TestNode, renters map[*TestNode]struct{}, hosts map[*TestNode]struct{}) error {
+	expectedContracts := defaultAllowance.Hosts
+	if uint64(len(hosts)) < expectedContracts {
+		expectedContracts = uint64(len(hosts))
+	}
+	for renter := range renters {
+		err := build.Retry(100, time.Second, func() error {
```

Comment:
No need to wait a full second between tries, can wait 100ms

Reply:
I would recommend to leave it at 1 second and add a todo. Forming contracts is incredibly slow at the moment. If we change it to 100 milliseconds, 100 tries are not enough. That is not a bad thing usually but we might not want to mine >100 blocks while waiting for the renters to form contracts.